notebooks/2018-12-14_Fortran_Files.ipynb

```diff
@@ -0,0 +1,333 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `np.fromfile` method reads the data stream in as-is, and iterprets the values according to the datatype you gave it, in the above case little-endian 2-byte integer.\n",
+    "\n",
+    "For an overview of possible data types, see [here](https://docs.scipy.org/doc/numpy/reference/arrays.interface.html#python-side).\n",
+    "\n",
+    "The integer values are correctly read in, but of course the 'H' and 'e' get mashed into a single integer value of 25928, 'll' becomes 27759, and so forth.\n",
+    "\n",
+    "Still, this might be the simplest way to transfer a single array bit-correct to python."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Unformatted sequential\n",
+    "\n",
+    "There is no standardised method to store unformatted sequential data, and the exact format might vary between different compilers and platforms.\n",
+    "\n",
+    "That said, most compilers seem to store it in a similar way by now.\n",
+    "\n",
+    "### Fortran Write\n",
+    "\n",
+    "```fortran\n",
+    "program write_unformatted\n",
+    "    use iso_fortran_env\n",
+    "    implicit none\n",
+    "    integer(kind=int16) :: ii\n",
+    "    integer :: output_handle\n",
+    "    open(newunit=output_handle, file='unformatted_data.dat', form='unformatted', &\n",
+    "        status='replace', action='write', access='sequential')\n",
+    "    write(output_handle) [(ii, ii=1, 10)]\n",
+    "    write(output_handle) \"Hello World\"\n",
+    "    close(output_handle)\n",
+    "end program write_unformatted\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "00000000  14 00 00 00 01 00 02 00  03 00 04 00 05 00 06 00  |................|\r\n",
+      "00000010  07 00 08 00 09 00 0a 00  14 00 00 00 0b 00 00 00  |................|\r\n",
+      "00000020  48 65 6c 6c 6f 20 57 6f  72 6c 64 0b 00 00 00     |Hello World....|\r\n",
+      "0000002f\r\n"
+     ]
+    }
+   ],
+   "source": [
+    "!hexdump -C unformatted_data.dat"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can still see the values of 1 through 10 (`01 00` through `0a 00`), but you can also see that it's no longer the first value. \n",
+    "It starts with `14 00 00 00`, or 20, which is the number of bytes that make this list up. After the array, the 20 is repeated, in case you read in reverse.\n",
+    "\n",
+    "Next comes `0b 00 00 00`, or 11 -- exactly the number of bytes in \"Hello World\", again followed by a repeat of the record header 11.\n",
+    "\n",
+    "### Python read\n",
+    "\n",
+    "To read this data in Python, you need to know the data type of the header, almost always an unsigned int, and usually 4 bytes in length:"
```

Comment:
Can be a 32 or 64 bit value, depending on compiler options. Also be careful about big vs little endian values. It may be useful to write a known header at the start of the file so this can be identified, but then if you're going that in depth you should be using netcdf